ext/ddtrace_profiling_native_extension/heap_recorder.c

```diff
@@ -0,0 +1,489 @@
+      if (queued_sample->obj == obj) {
+        queued_sample->skip = true;
+        break;
+      }
+    }
+
+    // free of an untracked object, return early
+    return;
+  }
+
+  // if we got this far, we freed a tracked object so need to update records!
+  int error = pthread_mutex_trylock(&heap_recorder->records_mutex);
+  if (error) {
+    // We weren't able to get a lock, so enqueue this sample for later processing
+    // and exit early
+    if (error == EBUSY) {
+      enqueue_free(heap_recorder, obj);
+    } else {
+      ENFORCE_SUCCESS_GVL(error)
+    }
+    return;
+  }
+
+  // If we got this far, we got a write lock so we can commit the record
+  commit_free(heap_recorder, obj);
+
+  pthread_mutex_unlock(&heap_recorder->records_mutex);
+}
+
+// ===============
+// Heap Record API
+// ===============
+heap_record* heap_record_init(const heap_stack *stack) {
+  heap_record* record = ruby_xcalloc(1, sizeof(heap_record));
+  record->num_tracked_objects = 0;
+  record->inuse_objects = 0;
+  record->stack = stack;
+  return record;
+}
+
+void heap_record_free(heap_record *record) {
+  ruby_xfree(record);
+}
+
+
+// =================
+// Object Record API
+// =================
+object_record* object_record_init(VALUE new_obj, unsigned int weight, heap_record *heap_record) {
+  object_record* record = ruby_xcalloc(1, sizeof(object_record));
+  record->weight = weight;
+  record->heap_record = heap_record;
+  return record;
+}
+
+void object_record_free(object_record *record) {
+  ruby_xfree(record);
+}
+
+// ==============
+// Heap Frame API
+// ==============
+int heap_frame_cmp(heap_frame *f1, heap_frame *f2) {
+  int cmp = strcmp(f1->name, f2->name);
+  if (cmp != 0) {
+    return cmp;
+  }
+  cmp = strcmp(f1->filename, f2->filename);
+  if (cmp != 0) {
+    return cmp;
+  }
+  return (int) (f1->line - f2->line);
+}
+
+st_index_t string_hash(char *str, st_index_t seed) {
+  return st_hash(str, strlen(str), seed);
+}
+
+st_index_t heap_frame_hash(heap_frame *frame, st_index_t seed) {
+  st_index_t hash = string_hash(frame->name, seed);
+  hash = string_hash(frame->filename, hash);
+  hash = st_hash(&frame->line, sizeof(frame->line), hash);
+  return hash;
+}
+
+// Important: This should match string_hash behaviour
+st_index_t char_slice_hash(ddog_CharSlice char_slice, st_index_t seed) {
+  return st_hash(char_slice.ptr, char_slice.len, seed);
+}
+
+// ==============
+// Heap Stack API
+// ==============
+heap_stack* heap_stack_init(ddog_prof_Slice_Location locations) {
+  heap_stack *stack = ruby_xcalloc(1, sizeof(heap_stack));
+  stack->frames = ruby_xcalloc(locations.len, sizeof(heap_frame));
+  stack->frames_len = locations.len;
+  for (uint64_t i = 0; i < locations.len; i++) {
+    const ddog_prof_Location *location = &locations.ptr[i];
+    heap_frame *frame = &stack->frames[i];
+    frame->name = ruby_strdup(location->function.name.ptr);
+    frame->filename = ruby_strdup(location->function.filename.ptr);
```

Comment:
There's a couple of assumptions here that I'm not 100% sure about:
* That `ptr` is always a null-terminated string: I think so... but... technically... we're getting this from a Ruby string, and I think it's possible for Ruby strings to not be null-terminated? Or at least I'm not entirely confident.
* That there's always a value in the `ptr`

When we feed the locations to libdatadog, we give it a pointer and a length (which I'm not sure is always non-zero) so neither of those problems would occur, but here since we're just using the pointers without the length I'm a bit more unsure.

(Note that it doesn't mean we need to change stuff here -- maybe we just test these things when building the locations so we don't have to care here)

Reply:
👍 

Adding a `ruby_strndup` helper to deal with this for now.